Fixed #25968 -- Changed project/app templates to use a "py-tpl" suffix.

lamby opened the pull request:
Debian packages unconditionally byte-compile .py files on installation and
do not silence errors by design. Therefore, we need a way of shipping these
invalid .py files without a .py extension but ensuring that when we
template them, they end up as .py.

We don't special-case .py files so that the all the TemplateCommand
command-line options (eg. extra_files and extensions) still work entirely
as expected and it may even be useful for other formats too.

Signed-off-by: Chris Lamb <chris@chris-lamb.co.uk>

timgraham commented:
[Florian suggests](https://groups.google.com/d/msg/django-developers/nPShxGatRuE/e4Wp0zC3BgAJ) that all Python template files use the suffix (even the valid ones). I'd suggest ".tpl" as the extension. A Trac ticket is also required.

lamby commented:
Updated - moved to ``.tpl``. Will need to check that this still means the files are included (without ``__init__.py`` tools like ``find_packages`` etc may not "find" them)

timgraham commented:
I think we need to document the behavior. This will be backwards incompatible for any custom templates that have files ending with ".tpl", right? Maybe there's some way to try to account for that, e.g. if a file has ".tpl" but not another extension?

claudep commented:
@lamby is there a ticket yet for this patch?

lamby commented:
@claudep There is now: https://code.djangoproject.com/ticket/25968

claudep commented:
Most if not all test failures can be fixed by:
```
diff --git a/tests/admin_scripts/tests.py b/tests/admin_scripts/tests.py
index 755fb0b..a37a71e 100644
--- a/tests/admin_scripts/tests.py
+++ b/tests/admin_scripts/tests.py
@@ -181,7 +181,7 @@ class AdminScriptTestCase(unittest.TestCase):
                 pass
 
         conf_dir = os.path.dirname(upath(conf.__file__))
-        template_manage_py = os.path.join(conf_dir, 'project_template', 'manage.py')
+        template_manage_py = os.path.join(conf_dir, 'project_template', 'manage.py.tpl')
 
         test_manage_py = os.path.join(self.test_dir, 'manage.py')
         shutil.copyfile(template_manage_py, test_manage_py)
```

lamby commented:
Updated, thanks :)

timgraham commented:
The branch needs a rebase and (in case you missed it) my comment is still outstanding, "I think we need to document the behavior. This will be backwards incompatible for any custom templates that have files ending with ".tpl", right? Maybe there's some way to try to account for that, e.g. if a file has ".tpl" but not another extension?"

lamby commented:
Rebased. Documentation to follow.

lamby commented:
Rebased & documentation added.

timgraham commented:
There's a remaining test failure: `project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers`

timgraham commented:
I think Claude's suggestion to limit the potential for backwards-incompatibilities is a good one:
What about limiting chopping to ".py.tpl" => ".py"?

lamby commented:
> What about limiting chopping to ".py.tpl" => ".py"?

Makes sense. Have rewritten the patch a little - by limiting it to ``.py`` it actually becomes more of a "search/replace" than simply a "stripping" one (or we require ugly special-casing of ``.py``). Docs updated to match.

(Test passing locally, will see what CI says...)

timgraham commented:
The failing test only runs on Python 3: `project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers`

Please check docs and flake8 failures too.

lamby commented:
> The failing test only runs on Python 3: project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers

Fixed.

> Please check docs and flake8 failures too.

Fixed.

claudep commented on docs/releases/1.9.2.txt:
```diff
@@ -21,3 +23,28 @@ Bugfixes
   the admin calendar widget for timezones from GMT+0100 to GMT+1200
   (:ticket:`24980`).
+* Fixed a regression in ``ConditionalGetMiddleware`` causing ``If-None-Match`` checks
+  to always return HTTP 200 (:ticket:`26024`).
```
That's probably a merge artifact. Please simply remove those 2 lines (currently duplicated).

lamby replied:
Updated

claudep commented on docs/releases/1.9.2.txt:
```diff
@@ -21,3 +23,28 @@ Bugfixes
+This caused difficulties for packaging systems that unconditionally
+byte-compile ``*.py`` files, either failing or generating confusing warnings.
+
+To overcome this, a ``.tpl`` suffix has been added to these Python files  so
```
double space between "files" and "so"

lamby replied:
Updated.

lamby commented:
Updated to match in-line comments.

timgraham commented on docs/releases/1.9.2.txt:
```diff
@@ -21,3 +23,26 @@ Bugfixes
+``.py.tpl`` suffix is also replaced with ``.py`` upon rendering - for example,
+a template with the filename ``manage.py.tpl`` will be saved at ``manage.py``.
+
+If you are already using custom Python templates that end with ``.py.tpl``, in
```
This advice doesn't work as `filename.py.tpl.tpl` wouldn't be rewritten.

timgraham commented:
Some edits: http://dpaste.com/3RKKDMH

lamby commented:
@timgraham Thanks! All merged.

timgraham commented:
Any response to the latest post on the mailing list thread about using py-tpl as a suffix?

lamby commented:
> Any response to the latest post on the mailing list thread about using py-tpl as a suffix?

(just to be clear, the code now does that)

timgraham commented:
I'm planning 1.9.2 for Feb. 1 so it would be great to get a consensus on that last question before then.

claudep commented:
@lamby `.py.tpl` vs `.py-tpl`

lamby commented:
@claudep @timgraham I like ``.py-tpl`` - it's immediately obvious that "something interesting" is going on, whilst it is possible for one to not notice ``.py.tpl``.

Shall I go ahead and update the patch?

timgraham commented:
I'm in favor.

lamby commented:
@timgraham @claudep Updated

timgraham commented:
merged in abc0777b63057e2ff97eee2ff184356051e14c47, thanks!
